Created a tech spec for Embedded Readmes

advay26-zz opened the pull request:
Created an implementation spec for the Embedded Readmes feature [[Issue 9752](https://github.com/NuGet/Home/issues/9752)].

advay26-zz committed:
Created blank tech spec file for Embedded Readmes

advay26-zz committed:
Completed first draft of spec

dominoFire commented on designs/EmbeddedReadmesTechSpec.md:
```diff
@@ -0,0 +1,102 @@
+
+## Non-Goals
+
+* Server side functionality
```
Include as a non-goal: parse readme as a markdown, because:
- we don't have built-in libraries to parse markdown, and adding one would take some time
- There's multiple Markdown implementations (github, daring fireball etc.)

nkolev92 commented on designs/EmbeddedReadmesTechSpec.md:
```diff
@@ -0,0 +1,102 @@
+
+**If no Readme property is present in the source file, but a readme.md file exists in the base directory**
+
+Once the properties and files have been processed, if the PackageBuilder object's Readme property is null, we need to look through the base directory for a "readme.md" file. If such a file exists, we must include it in the package and and populate the Readme property with "readme.md".
```
I'd caution against this.

There's no precedent to this approach, and there are features with similar behavior, see licenses, icons. 

Users being explicit about metadata is good!

nkolev92 replied:
There are really 3 pack scenarios. 

1. dotnet pack(msbuild pack). All the content that needs included in the package will be included through Content/None items through MSBuild. We should not do additional searches. We should not pull in things accidentally that might not be part of the project. 
2. nuget.exe pack with a nuspec - without a files element...this basically will include all files under the nuspec directory.
3. nuget.exe pack with a nuspec - with a files element - we will include only the files that the files element has specified. 

In neither of the above 3 cases, do we arbitrarily glob looking for a file to include as metadata.

advay26-zz replied:
I've removed this functionality, and the references to it within this spec, based on feedback from the review meeting and Chris' interviews with customers.

nkolev92 commented on designs/EmbeddedReadmesTechSpec.md:
```diff
@@ -0,0 +1,102 @@
+public void CheckForReadme(string basePath)
+{
+    string readmeName = "readme.md";
+    List<PhysicalPackageFile> searchFiles = ResolveSearchPattern(basePath, readmeName, string.Empty, false).ToList();
```
We already glob a couple of times during pack, we don't need an additional glob.

advay26-zz committed:
Added a non-goal: "parsing readme as markdown"

nkolev92 commented on designs/EmbeddedReadmesTechSpec.md:
```diff
@@ -0,0 +1,102 @@
+To validate the Readme property and the corresponding file when `pack` is called, the ValidateReadmeFile method in [PackageBuilder.cs](https://github.com/NuGet/NuGet.Client/blob/0f8ad8263539cb9bc69c441569453c1da98fb4cc/src/NuGet.Core/NuGet.Packaging/PackageCreation/Authoring/PackageBuilder.cs) checks that:
+* The file extension matches a Markdown file
+* The file specified in the Readme property exists in the package's files
+* The file can be opened
```
Can you clarify what opening a file means in this case?

advay26-zz replied:
This was meant to cover any other exceptions thrown while attempting to open the embedded readme file, but I've now removed it from the spec in order to be more clear about the validation goals.

nkolev92 commented on designs/EmbeddedReadmesTechSpec.md:
```diff
@@ -0,0 +1,102 @@
+
+## Open Questions
+
+* Should we warn users if the Readme property is empty?
```
If they don't have a readme at all? 
I'd vote no for now. 
Lots of customer elevate warnings to errors and while a good quality package will have a readme, I'd rather that get added during the actual high quality package effort (@chgill-MSFT, @JonDouglas will know more).

nkolev92 reviewed (commented):
This is a great write-up @advay26. 
It's very detailed. Given that I'm not gonna be able to make the meeting, adding my comments now.

advay26-zz committed:
Updated the spec in response to feedback

Updated the tech spec based on feedback from the review meeting and Customer input

dominoFire commented on designs/EmbeddedReadmesTechSpec.md:
```diff
@@ -0,0 +1,85 @@
+
+## Solution
+
+**NOTE:** To see the code described in this section, you can look at the edited files [here](https://github.com/NuGet/NuGet.Client/compare/dev-advay26-readme).
```
Will the link be alive once this spec is merged? Right now it points to a new page.

advay26-zz replied:
I've changed the link to point to my (Draft) PR instead of the branch I was working on so that the link will still be functional  even once the PR is merged and the branch is deleted.

dominoFire commented on designs/EmbeddedReadmesTechSpec.md:
```diff
@@ -0,0 +1,85 @@
+## Goals
+
+* Implement the Client side functionality for Embedded Readmes:
+    * nuget pack support
+    * dotnet pack support
+    * Validation of readme files
```
Add a bullet point for MSBuild, since you're already adding support in the NuGet Pack targets.

```suggestion
    * `nuget pack` support
    * `dotnet pack` support
    * `msbuild /t:pack` support
    * Validation of readme files
```

dominoFire commented on designs/EmbeddedReadmesTechSpec.md:
````diff
@@ -0,0 +1,85 @@
+```
+AddElementIfNotNull(elem, ns, "readme", metadata.Readme);
+```
+
````
Describe briefly how are you reading the new readme property. Tip: Look at NuspecReader class

advay26-zz replied:
I've added a reference to how the Readme property is read by the `GetReadme` method in the `NuspecReader` class. I imagine this is what you were referring to, but feel free to correct me if I'm wrong.

dominoFire reviewed (approved):
Minor comments

advay26-zz committed:
Added MSBuild pack to goals, referenced NuspecReader work

dominoFire commented on designs/EmbeddedReadmesTechSpec.md:
```diff
@@ -0,0 +1,88 @@
+
+## Problem Background
+
+Users of NuGet want to be able to include Readme files as part of the package when they `pack` and `push` their project through the commandline. The [design document](https://github.com/NuGet/Home/wiki/Packaging-READMEs-within-the-nupkg) outlines the work needed for this feature and the intended results.
```
The link to `https://github.com/NuGet/Home/wiki/Packaging-READMEs-within-the-nupkg` points to a new wiki page. Is it intentional? Are you going to create a new wiki page once this PR is merged?

advay26-zz replied:
It looks like the name of the document within the wiki was changed at some point. I've updated the link to point to the correct place now. Thanks for pointing this out!

advay26-zz committed:
Updated the link to the design document

advay26 commented:
I've made changes to the tech spec based on feedback I've received, specifically addressing concerns about the automatic packing of READMEs (now disabled) and the Validation strategy (who validates the actual markdown file). I wasn't able to request reviews from some of you through the Reviewers tab, so I'm tagging you here to request comments/feedback/approvals.

@zkat @zivkan @sbanni @erdembayar @rrelyea @srdjanjovcic @cristinamanum @heng-liu @donnie-msft

donnie-msft commented on designs/EmbeddedReadmesTechSpec.md:
```diff
@@ -0,0 +1,88 @@
+To validate the Readme property and the corresponding file when `pack` is called, the ValidateReadmeFile method in [PackageBuilder.cs](https://github.com/NuGet/NuGet.Client/blob/0f8ad8263539cb9bc69c441569453c1da98fb4cc/src/NuGet.Core/NuGet.Packaging/PackageCreation/Authoring/PackageBuilder.cs) checks that:
+* The file extension matches a Markdown file
+* The file specified in the Readme property exists in the package's files
+* The file size does not exceed 1 MB
```
These validations are mainly for nuget.org, right? If we eventually want to have looser restrictions for readmes for local or private feeds, we feel this is a minimum set of validations going forward? eg, In future iterations of this feature, I could have a 5GB readme on my private feed... who cares? But I can't publish to nuget.org with that. 

I think the answer here is "yes", so may want to mention that in the description how we feel this validation set could evolve without causing backwards compatibility issues.

If you don't like the idea of having Pack validation eventually removed/less restrictive, then I'd question why we need to validate on Pack and not on Push?...
```
Pack -> warn that nuget.org won't accept >1MB
Push -> error from nuget.org (server doesn't accept > 1MB)
(in the future)
Push  (NOT nuget.org) -> accept/error depending on server behavior
Push (local feed) -> local feed doesn't care
```

advay26-zz replied:
@chgill-MSFT I was hoping you could clarify this

chrisraygill replied:
This is a good question!

Advay and I had a chat about this. It looks like we don't validate for size on license file today. In general we don't make any assumptions on pack that a package isn't meant for NuGet.org today. Not only that, but the likelihood that a README will be larger than 1mb is very low.

Ultimately, we came to the conclusion that is makes sense to remove the size check on pack and only validate README file size on push/upload to NuGet.org.

donnie-msft replied:
Sounds good! I think keeps our options open in the long run. 
Are we planning on changing `pack` to `push` in this paragraph?

donnie-msft replied:
Nevermind, yes the basic checks will remain on `pack`. I'm guessing `push` validations are out of scope? May want to indicate that in **Future**?

advay26-zz replied:
Since we won't be doing the validation on our side and the size check will only be done by nuget.org, should we say we validate size on `push`, or should we just add a note saying that "NuGet.org will only accept README files that do not exceed 1 MB"?

advay26-zz replied:
I'll add the 1 MB size validation to the "Considerations" section and explain why we won't be doing it here. I'll include a reference to NuGet.org's 1 MB size limit there, and mention that this would result in an error when someone uses `push` to upload a package to NuGet.org. Does that seem fair?

advay26-zz replied:
Done!

donnie-msft commented on designs/EmbeddedReadmesTechSpec.md:
```diff
@@ -0,0 +1,88 @@
+
+* Are there any other validation checks to do when packing a Readme?
+
+## Considerations
```
I agree with these. It's opt-in for now.

donnie-msft commented on designs/EmbeddedReadmesTechSpec.md:
```diff
@@ -0,0 +1,88 @@
+
+**NOTE:** The Server team will parse and validate the markdown contents of the Readme file when it is uploaded.
+
+## Future Work
```
👍

donnie-msft commented on designs/EmbeddedReadmesTechSpec.md:
```diff
@@ -0,0 +1,88 @@
+    * `msbuild /t:pack` support
+    * Validation of readme files
+
+## Non-Goals
```
May want to make it clear that this excludes the Visual Studio property page selection of a Readme.

advay26-zz committed:
Clarified VS PP UI's Readme support as a non-goal here

advay26-zz committed:
Updated the PR that the code hyperlink points to

advay26-zz committed:
Removed 1MB size check for Readme files

advay26-zz committed:
Included the 1 MB size limit in the 'Considerations' section

donnie-msft reviewed (approved):
Awesome! 👏

JonDouglas commented:
**Open Questions**
Are there any other validation checks to do when packing a Readme?

> When a user packs a NuGet package without a README, there should be a warning.

**Considerations**
1. Should we warn users if the Readme property is empty?
Based on feedback from both the NuGet team and customers, we should not do this. The goal of encouraging user adoption of Embedded Readmes can be pursued as part of a separate Package Quality endeavor.

> I believe we should set a standard with this feature today. READMEs are a critical component for someone to not have to navigate away from NuGet.org or Visual Studio in the sense that they can learn about the package & get started with it quickly. Many customers said they would prefer to not have to go to GitHub to view the README if possible.

2. Automatically packing a Readme file from the base directory
Similar to the above point, this functionality is not something that we will look to add right now. Relying on explicit user input for the metadata is more desirable, and avoids springing any surprises on customers.

> Does this mean adding a README by default? I don't think we need to do that for the scope of this spec.

3. Enforcing a size limit on Readme files on pack
Originally, we intended to enforce a 1 MB size limit on Embedded Readme files both while running pack on the Client-side and while using push to upload to NuGet.org. However, since we do not want to impose such size restrictions on users looking to upload packages to private or local feeds, we decided against throwing an error for this on pack. Now, the user can run pack to create packages with Readme files over 1 MB, but will receive an error if they try to upload such a package to NuGet.org, which enforces a strict 1 MB limit.

> The follow-up question here is that most readmes should generally be in the KB range & all resources for the README is hosted somewhere else like github content or another location. Will embedded readmes support this similar concept where outside resources can be accessed or will users need to embed those resources as well?

chgill-MSFT commented:
I followed up with Jon in chat regarding the above, but here is the response I gave for transparency:

>1.  I believe we should set a standard with this feature today. READMEs are a critical component for someone to not have to navigate away from NuGet.org or Visual Studio in the sense that they can learn about the package & get started with it quickly. Many customers said they would prefer to not have to go to GitHub to view the README if possible.

As @zivkan has brought in past discussions, warnings can be problematic for some customers who elevate warnings to errors. Since a README isn't strictly necessary for private/internal package use, this could be a frustrating experience for those who don't need it. 

The alternative is that we just print a "suggestion" or message as this would increase feature awareness without being overly aggressive. I decided not to do this for the initial release for a few reasons:
1. Customer feedback indicated interest in an overall package quality auditing experience which is aligned with @JonDouglas package quality indicators work. "Suggestions" will likely be apart of that work.
1. Our README support on NuGet.org is not at full parity with GitHub and may have some friction for customers to get their READMEs "NuGet.org ready."  We should understand and remove these friction points as well as we can before pinging customers for not adopting READMEs.
1. Most client/pack validations today are private/public package agnostic. Customer's who are creating private packages that don't need READMEs have no way of disabling these suggestions.

> 2. Does this mean adding a README by default? I don't think we need to do that for the scope of this spec.

👍

> 3. The follow-up question here is that most readmes should generally be in the KB range & all resources for the README is hosted somewhere else like github content or another location. Will embedded readmes support this similar concept where outside resources can be accessed or will users need to embed those resources as well?

We will not support embedded resources for README images in this initial release. Allowlist --> proxy service is a more likely evolution, so size shouldn't be much of an issue. 1mb is a lot for text.

nkolev92 commented on designs/EmbeddedReadmesTechSpec.md:
```diff
@@ -0,0 +1,93 @@
+
+## Future Work
+
+* Adding readme support through the VS Project Properties UI (Project System)
```
It might be nice to provide links to the future work.

advay26-zz replied:
Done!

nkolev92 reviewed (approved):
This is one of the most detailed technical specs I have seen in a while :) 
Looks great @advay26!

advay26-zz committed:
Added links to the tracking issues for the 'Future Work' tasks
